lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,392 @@
+use crate::TxSyncError;
+
+use lightning::util::logger::Logger;
+use lightning::{log_error, log_given_level, log_info, log_internal, log_debug, log_trace};
+use lightning::chain::WatchedOutput;
+use lightning::chain::{Confirm, Filter};
+
+use bitcoin::{BlockHash, BlockHeader, Script, Transaction, Txid};
+
+use esplora_client::Builder;
+#[cfg(feature = "async-interface")]
+use esplora_client::r#async::AsyncClient;
+#[cfg(not(feature = "async-interface"))]
+use esplora_client::blocking::BlockingClient;
+
+use std::sync::Mutex;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::collections::HashSet;
+use core::ops::Deref;
+
+/// Synchronizes LDK with a given [`Esplora`] server.
+///
+/// Needs to be registerd with a [`ChainMonitor`] via the [`Filter`] interface to be informed of
+/// transactions and outputs to montor.
+///
+/// [`Esplora`]: https://github.com/Blockstream/electrs
+/// [`ChainMonitor`]: lightning::chain::chainmonitor::ChainMonitor
+/// [`Filter`]: lightning::chain::Filter
+pub struct EsploraSyncClient<L: Deref>
+where
+	L::Target: Logger,
+{
+	// Transactions that were registered via the `Filter` interface and have to be processed.
+	queued_transactions: Mutex<HashSet<Txid>>,
+	// Transactions that were previously processed, but must not be forgotten yet.
+	watched_transactions: Mutex<HashSet<Txid>>,
+	// Outputs that were registered via the `Filter` interface and have to be processed.
+	queued_outputs: Mutex<HashSet<WatchedOutput>>,
+	// Outputs that were previously processed, but must not be forgotten yet.
+	watched_outputs: Mutex<HashSet<WatchedOutput>>,
+	// Indicates whether we need to resync, e.g., after encountering an error.
+	pending_sync: AtomicBool,
+	// The tip hash observed during our last sync.
+	#[cfg(not(feature = "async-interface"))]
+	last_sync_hash: std::sync::Mutex<Option<BlockHash>>,
+	#[cfg(feature = "async-interface")]
+	last_sync_hash: futures::lock::Mutex<Option<BlockHash>>,
```

Comment:
Hm, we could split the queues out in a separate struct, but I'm not sure if that's preferable, since we'd then need to lock the entire thing when registering via `register_tx`/`register_output`. Moreover, I don't think we can have the remaining fields split out under a single lock.

Reply:
> Why is that? Ah, is it because we can't assume `EsploraSyncClient::sync` won't be called concurrently on two threads?

Well I *thought* it wouldn't be possible since we needed to access some of the other fields such as `pending_sync` from another thread, yes. However, this was only the case in a prior interation of the code. While trying to answer your question I realized all reasons not to do that are basically gone by now. I now refactored the codebase quite a bit with https://github.com/lightningdevkit/rust-lightning/pull/1870/commits/dbb4199bb4e649cea07bf31dc0889ab472bce4ed and split out `SyncState` and `FilterQueue` which allows us to reduce the number of `Mutex`es to two. Thanks! :)
 
> I wonder if we had a queue abstraction that has internal locking and implements `chain::Filter`, then we could give this to both `ChainMonitor` and `EsploraSyncClient`, where the latter pops from the queue in order to populate `watched_transactions` and `watched_outputs`. Then `EsploraSyncClient::sync` can take `&mut self` and we can drop the locks. Let me know if I'm missing something.

Yes, as mentioned above I now introduced such an abstraction. This also has the added benefit that I'll be able to DRY up my code when introducing the same functionality in the Electrum version. However I'm neither a big fan of exposing this additional object in the interface nor the idea of exposing `&mut` in an API, as it just hotpotatoes type complexity to the user and potentially even limits how it can be used. I think internal mutability is much more preferable here?!

Reply:
> However I'm neither a big fan of exposing this additional object in the interface nor the idea of exposing &mut in an API, as it just hotpotatoes type complexity to the user and potentially even limits how it can be used. I think internal mutability is much more preferable here?!

Just want to +1 here for interior mutability; BDK has some [issues](https://github.com/bitcoindevkit/bdk/issues/165) which require us wrap their `bdk::Wallet` in a `tokio::sync::Mutex`, which then causes all sorts of problems like us having to change normally synchronous methods on our wallet wrapper (like `get_new_address`) into asynchronous ones because it's too dangerous to use `Mutex::try_lock()` while attempting to avoid deadlocks / flakiness. Changing the API to take `&mut self` would cause similar kinds of problems for us, since `sync()` is an (async) function that takes a long time and therefore the `Mutex` we use to wrap `EsploraSyncClient` must be async. Alternatively we'd have to allow only one handle to the sync client at one time, which complicates use cases like having a sync step in the init code and then moving the one and only client into a task for periodic resync. Interior mutability should be preferred, even if it does make the code a bit uglier in order to support it.

Reply:
Current approach with fewer locks, sub-structs, and interior mutability SGTM.